Unit Test: ft_on_transfer

bh2smith opened the pull request:
We had already unit tested this, but now we uncomment the remaining block and test the happy path.

bh2smith committed:
Upgrade Dependencies & Delete Draft (#1)

Upgrades all the contract dependencies.

Deletes Draft (since its undocumented what it is for an not obvious).

Started adding tests (easy)

Removed e2e test directory (will add back later).

bh2smith committed:
unit test ft-receiver

frol commented on src/callbacks.rs:
```diff
@@ -36,61 +43,62 @@ impl SelfCallbacks for Contract {
                     remove_indices.push(index);
                 }
-                total_balance += claim_amount.0;
+                total_balance = total_balance.saturating_add(claim_amount);
                 let event = FtLockupClaimLockup {
                     id: index,
                     amount: claim_amount,
                 };
                 events.push(event);
```
This could lead to confusion or vulnerability. Let's use `checked_*` instead of `saturating_*`

frol commented on src/callbacks.rs:
```diff
@@ -36,61 +43,62 @@ impl SelfCallbacks for Contract {
                 self.internal_save_account_lockups(&account_id, indices);
             }
-            emit(EventKind::FtLockupClaimLockup(events));
+            // TODO: Should we emit a single vector or all separate?
+            events.iter().for_each(|event| event.emit());
```
Firing individual logs vs batching is a matter of gas usage. If I remember correctly from other testing, we can fire up to 300 events with 300TGas (max limit).

How many events do we expect here? Why not batching?

frol commented on src/lib.rs:
```diff
@@ -24,138 +20,66 @@ pub mod termination;
-    pub deposit_whitelist: UnorderedSet<AccountId>,
-
-    /// account ids that can perform all actions on drafts:
-    /// - manage drafts, draft_groups
-    pub draft_operators_whitelist: UnorderedSet<AccountId>,
-
-    pub next_draft_id: DraftIndex,
-    pub drafts: LookupMap<DraftIndex, Draft>,
-    pub next_draft_group_id: DraftGroupIndex,
-    pub draft_groups: UnorderedMap<DraftGroupIndex, DraftGroup>,
+    /// - manage deposit_allowlist
+    /// - create lockups, terminate lockups
+    pub deposit_allowlist: UnorderedSet<AccountId>,
```
I want to highlight for anyone reviewing this code in the future that there is a breaking change in the STATE layout, so this version of the contract cannot be deployed over the old version of the contract without a migration

frol commented on src/lib.rs:
```diff
@@ -24,138 +20,66 @@ pub mod termination;
-    /// - create lockups, terminate lockups, fund draft_groups
-    pub deposit_whitelist: UnorderedSet<AccountId>,
-
-    /// account ids that can perform all actions on drafts:
-    /// - manage drafts, draft_groups
-    pub draft_operators_whitelist: UnorderedSet<AccountId>,
-
-    pub next_draft_id: DraftIndex,
-    pub drafts: LookupMap<DraftIndex, Draft>,
-    pub next_draft_group_id: DraftGroupIndex,
-    pub draft_groups: UnorderedMap<DraftGroupIndex, DraftGroup>,
```
It is more than unit-tests addition and SDK migration. Could you shed some light on the context you have? Looking at this code now I tend to agree that we should have considered implementing draft functionality as a separate contract or even off-chain, but I wonder what is your plan and how you envision the usage of this contract.

bh2smith commented:
Hey @frol - sorry this PR was meant to be targeted at my fork. I would never intentionally propose so many changes without context. 

Since we are here - this is my approach:

1. **Upgrade all dependencies:** You'll notice that I removed drafting here because the project is not very well documented and I couldn't tell what it was for at the time (wanted to hone in on the rest). I also removed the e2e tests here with intention of adding them back using the latest dependencies. This was https://github.com/bh2smith/ft-lockup-contract/pull/1
2. **Add unit tests** - to confirm my own understanding of the undocumented code. This will occur over several PRs.
3. **Add back the e2e tests** - as they were with latest dependencies.
4. Once the whole thing is better I might add back draft. Would love some insight on what it is for. 

I can close this PR and point it at my own fork for now - I was going to come back to you with questions, but haven't had so many yet. 


BTW - this project is very well put together. Thanks for making this easy to follow and having a nice clean codebase 🥳

frol commented:
@bh2smith  thanks for the detailed explanation! I’m looking forward to see this contract widely used. Please, keep us posted and feel free to skip drafts functionality, as I said, likely, we’d better extract it into a separate service, so this contract can be kept easily auditable and hopefully immutable for years.
